File: 2. addTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        //create a dummy node
        ListNode *dummy=new ListNode(0), *p=dummy;
        int flag=0;
        while(l1 || l2 || flag){
            if(l1){
                flag+=l1->val;
                l1=l1->next;
            }
            if(l2){
                flag+=l2->val;
                l2=l2->next;
            }
            p->next=new ListNode(flag%10);
            flag=flag/10;
            p=p->next;
        }
        return dummy->next;
    }
};

File: 2.addTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if l1==None:
            return l2
        if l2==None:
            return l1
        dummy=ListNode(-1)
        p=dummy
        flag=0
        while l1 and l2:
            p.next=ListNode((l1.val+l2.val+flag)%10)
            flag=(l1.val+l2.val+flag)//10
            l1=l1.next
            l2=l2.next
            p=p.next
        if l1:
            while l1:
                p.next= ListNode((l1.val + flag) % 10)
                flag = (l1.val + flag) // 10
                l1 = l1.next
                p = p.next
        if l2:
            while l2:
                p.next=ListNode((l2.val+flag)%10)
                flag=(l2.val+flag)//10
                l2=l2.next
                p=p.next
        if flag==1:
            p.next=ListNode(1)
        return dummy.next
